src/shared/Microsoft.Git.CredentialManager/Interop/Windows/WindowsCredentialManager.cs

```diff
@@ -9,119 +11,289 @@ namespace Microsoft.Git.CredentialManager.Interop.Windows
-                TargetName = key,
-                CredentialBlob = Marshal.AllocHGlobal(passwordBytes.Length),
-                CredentialBlobSize = passwordBytes.Length,
-                Persist = CredentialPersist.LocalMachine,
-                AttributeCount = 0,
-                UserName = credential.UserName,
-            };
+                service = service.TrimUntilIndexOf(_namespace);
+            }
 
-            try
+            return new WindowsCredential(service, credential.UserName, password, credential.TargetName);
+        }
+
+        private bool IsMatch(string service, string account, Win32Credential credential)
+        {
+            // Match against the username first
+            if (!string.IsNullOrWhiteSpace(account) &&
+                !StringComparer.Ordinal.Equals(account, credential.UserName))
             {
-                Marshal.Copy(passwordBytes, 0, w32Credential.CredentialBlob, passwordBytes.Length);
+                return false;
+            }
 
-                int result = Win32Error.GetLastError(
-                    Advapi32.CredWrite(ref w32Credential, 0)
-                );
+            // Trim the "LegacyGeneric" prefix Windows adds and any namespace we have been filtered with
+            string targetName = credential.TargetName.TrimUntilIndexOf(TargetNameLegacyGenericPrefix);
+            if (!string.IsNullOrWhiteSpace(_namespace))
+            {
+                targetName = targetName.TrimUntilIndexOf($"{_namespace}:");
+            }
 
-                Win32Error.ThrowIfError(result, "Failed to write item to store.");
+            // If the target name matches the service name exactly then return 'match'
+            if (StringComparer.Ordinal.Equals(service, targetName))
+            {
+                return true;
             }
-            finally
+
+            // Try matching the target and service as URIs
+            if (Uri.TryCreate(service, UriKind.Absolute, out Uri serviceUri) &&
+                Uri.TryCreate(targetName, UriKind.Absolute, out Uri targetUri))
             {
-                if (w32Credential.CredentialBlob != IntPtr.Zero)
+                // Match host name
+                if (!StringComparer.OrdinalIgnoreCase.Equals(serviceUri.Host, targetUri.Host))
+                {
+                    return false;
+                }
+
+                // Match port number
+                if (!serviceUri.IsDefaultPort && serviceUri.Port == targetUri.Port)
+                {
+                    return false;
+                }
+
+                // Match path
+                if (!string.IsNullOrWhiteSpace(serviceUri.AbsolutePath) &&
+                    !StringComparer.OrdinalIgnoreCase.Equals(serviceUri.AbsolutePath, targetUri.AbsolutePath))
                 {
-                    Marshal.FreeHGlobal(w32Credential.CredentialBlob);
+                    return false;
                 }
+
+                // URLs match
+                return true;
             }
```

Comment:
This part is also not covered.

Reply:
This part is now exercised by the tests that creates two credentials with the same service name (so the target name will not directly equal the service name, since the target now also contains the username).